core/src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -180,131 +266,55 @@ public static Script parse(XContentParser parser, ParseFieldMatcher matcher, Str
-        } else if (token != Token.START_OBJECT) {
-            throw new ParsingException(parser.getTokenLocation(),
-                "unexpected value [" + (token.isValue() ? parser.text() : token) + "], expected [{, <code>]");
         }
 
-        ScriptType type = null;
-        String idOrCode = null;
-        Map<String, String> options = new HashMap<>();
-        Map<String, Object> params = null;
-
-        String name = null;
-
-        while ((token = parser.nextToken()) != Token.END_OBJECT) {
-            if (token == Token.FIELD_NAME) {
-                name = parser.currentName();
-            } else if (matcher.match(name, ScriptType.INLINE.getParseField())) {
-                if (type != null) {
-                    throw new ParsingException(parser.getTokenLocation(),
-                        "unexpected script type [" + ScriptType.INLINE.getParseField().getPreferredName() + "], " +
-                        "when type has already been specified [" + type.getName() + "]");
-                }
-
-                type = ScriptType.INLINE;
-
-                options = new HashMap<>();
-
-                if (parser.currentToken() == Token.START_OBJECT) {
-                    options.put(CONTENT_TYPE_OPTION, parser.contentType().mediaType());
-                    XContentBuilder builder = XContentFactory.contentBuilder(parser.contentType());
-                    idOrCode = builder.copyCurrentStructure(parser).bytes().utf8ToString();
-                } else {
-                    idOrCode = parser.text();
-                }
-            } else if (matcher.match(name, ScriptType.STORED.getParseField())) {
-                if (type != null) {
-                    throw new ParsingException(parser.getTokenLocation(),
-                        "unexpected script type [" + ScriptType.STORED.getParseField().getPreferredName() + "], " +
-                            "when type has already been specified [" + type.getName() + "]");
-                }
-
-                type = ScriptType.STORED;
-
-                if (token == Token.VALUE_STRING) {
-                    idOrCode = parser.text();
-                } else {
-                    throw new ParsingException(parser.getTokenLocation(),
-                        "unexpected value [" + (token.isValue() ? parser.text() : token) + "], expected [<id>]");
-                }
-            } else if (matcher.match(name, ScriptType.FILE.getParseField())) {
-                if (type != null) {
-                    throw new ParsingException(parser.getTokenLocation(),
-                        "unexpected script type [" + ScriptType.FILE.getParseField().getPreferredName() + "], " +
-                            "when type has already been specified [" + type.getName() + "]");
-                }
+        return PARSER.apply(parser, () -> matcher, Collections.singletonMap(LANG_PARSE_FIELD.getPreferredName(), lang));
```

Comment:
I see why you did this with the map here. I wonder if this is more of an argument for a builder because you can set the lang on the builder before you do the parsing.